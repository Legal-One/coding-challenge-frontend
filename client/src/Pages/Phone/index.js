import React, { useState, useEffect } from "react";
import axios from "axios";
import NumberCallLogs from "../../components/NumberCallLogs";
import {useParams} from "react-router-dom";
import "./index.scss";

const apiUrl = "http://localhost:5000/api/call/";

function Phone() {
    const {phoneNumber} = useParams();

    const [phoneNumberCallLogs, setPhoneNumberCallLogs] = useState([]);

    useEffect(() => {
        const fetchPhoneNumberCallLogs = async () => {
            const response = await axios.get(apiUrl + phoneNumber);
            setPhoneNumberCallLogs(response.data);
        };
        fetchPhoneNumberCallLogs();
    }, [phoneNumber]);


    return (
        <div>
            <NumberCallLogs phoneNumberCallLogs={phoneNumberCallLogs}/>
        </div>
    );
}
export default Phone;