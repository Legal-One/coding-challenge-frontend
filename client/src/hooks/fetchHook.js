import { useState, useEffect } from "react";
import axios from "axios";

export const useFetchData = (apiUrl) => {
    const [data, setData] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [isError, setIsError] = useState(false);

    useEffect(() => {
        const fetchData = async () => {
            setIsError(false);
            setIsLoading(true);

            try {
                const result = await axios(apiUrl);
                setData(result.data);
            } catch (error) {
                setIsError(true);
            }

            setIsLoading(false);
        };

        fetchData();
    }, [apiUrl]);

    return [data, isLoading, isError];
};
