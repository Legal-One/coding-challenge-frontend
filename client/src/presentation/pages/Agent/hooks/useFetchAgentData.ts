import { useEffect, useState } from 'react';
import moment from 'moment';
import useSWR from 'swr';
import fetcher from '../../../../utils/fetcher';
import formatNumber from '../../../../utils/formatNumber';

import { AgentData } from '../types';
import { LogData } from '../../Dashboard/types';
import { BASE_API } from '../../../../constants';

const useFetchAgentData = (identifier: number) => {
	const { data: logData, error: logError } = useSWR(`${BASE_API}/agent/${identifier}`, fetcher);
	const [agentData, setAgentData] = useState<AgentData | null>(null);
	const [agentLogs, setAgentLogs] = useState<Array<LogData>>([]);

	if (logError) console.log(logError);
	useEffect(() => {
		if (logData) {
			const { agent, log: agentLogData } = logData.data;
			const logs = [...agentLogData].map(log => ({
				...log,
				id: log.identifier,
				agent: `${log.firstName} ${log.lastName}`,
				date: moment(log.dateTime).calendar(),
				time: moment(log.dateTime).format('LT'),
				duration: formatNumber(log.duration),
			}));
			setAgentLogs(logs);
			setAgentData(agent);
		}
	}, [logData]);

	return { agentData, agentLogs, logError };
};

export default useFetchAgentData;
