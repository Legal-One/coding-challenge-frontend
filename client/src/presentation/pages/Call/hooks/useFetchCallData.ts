import { useEffect, useState } from 'react';
import moment from 'moment';
import useSWR from 'swr';
import fetcher from '../../../../utils/fetcher';
import formatNumber from '../../../../utils/formatNumber';

import { BASE_API } from '../../../../constants';
import { LogsInterface } from '../types';

const useFetchCallData = (number: number) => {
	const { data: logData, error: logError } = useSWR(`${BASE_API}/call/${number}`, fetcher);
	const [callRecords, setCallRecords] = useState<Array<LogsInterface>>([]);

	if (logError) console.log(logError);
	useEffect(() => {
		if (logData) {
			const allLogRecords = [...logData.data].map(log => ({
				...log,
				id: log.identifier,
				agent: `${log.firstName} ${log.lastName}`,
				date: moment(log.dateTime).calendar(),
				time: moment(log.dateTime).format('LT'),
				duration: formatNumber(log.duration),
			}));
			setCallRecords(allLogRecords);
		}
	}, [logData]);

	return { callRecords, logError };
};

export default useFetchCallData;
