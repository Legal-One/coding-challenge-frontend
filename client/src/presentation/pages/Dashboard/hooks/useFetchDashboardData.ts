import { useEffect, useState } from 'react';
import moment from 'moment';
import useSWR from 'swr';
import fetcher from '../../../../utils/fetcher';
import formatNumber from '../../../../utils/formatNumber';

import { LogData, LogExtendedData, DashboardFetchHook } from '../types';
import { BASE_API } from '../../../../constants';

const useFetchDashboardData: DashboardFetchHook = () => {
	const { data: logData, error: logError } = useSWR(`${BASE_API}`, fetcher);
	const { data: agentsData, error: agentsError } = useSWR(`${BASE_API}/agents`, fetcher);
	const [dashboardData, setDashboardData] = useState<Array<LogExtendedData>>([]);
	const [allLogs, setAllLogs] = useState<Array<LogData>>([]);
	const [totalCalls, setTotalCalls] = useState<number>(0);
	const [totalCustomers, setTotalCustomers] = useState<number>(0);
	const [totalAgents, setTotalAgents] = useState<number>(0);

	if (logError || agentsError) console.log(logError | agentsError);
	useEffect(() => {
		if (logData) {
			const allLogData = [...logData.data].map(log => ({
				...log,
				id: log.identifier,
			}));

			const uniqueNumbers: string[] = [...new Set(allLogData.map(item => item.number))];
			const transformedLogs: LogExtendedData[] = uniqueNumbers.map(num => {
				const logNumber = allLogData.filter(log => log.number === num);
				const count = logNumber.length;
				const lastRecord = logNumber[count - 1];
				lastRecord.agent = `${lastRecord.firstName} ${lastRecord.lastName}`;
				lastRecord.date = moment(lastRecord.dateTime).calendar();
				lastRecord.time = moment(lastRecord.dateTime).format('LT');
				lastRecord.duration = formatNumber(lastRecord.duration);
				return { ...lastRecord, count };
			});
			setDashboardData(transformedLogs);
			setTotalCalls(allLogData.length);
			setTotalCustomers(transformedLogs.length);
			setAllLogs(allLogData);
		}
	}, [logData]);
	useEffect(() => {
		if (agentsData) {
			setTotalAgents(agentsData.data.length);
		}
	}, [agentsData]);

	return { allLogs, dashboardData, logError, agentsError, totalCalls, totalCustomers, totalAgents };
};

export default useFetchDashboardData;
