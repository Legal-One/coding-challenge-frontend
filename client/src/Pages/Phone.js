import React from "react";
import { useParams } from "react-router-dom";
import { useFetchData } from "../hooks/fetchHook";
import Table from "../components/Table";
import Chart from "../components/Chart";
import { getFormattedDateTime } from "../utils/timeFormat";
import "./index.scss";

const apiUrl = "http://localhost:5000/api/call/";
const headers = ["Agent Name", "Call date and time", "Resolution"];

const dataMapper = (phoneLogs) => {
    return phoneLogs.map((phoneLog) => [
        phoneLog.agentName,
        getFormattedDateTime(phoneLog.callTime),
        phoneLog.resolution,
    ]);
};

function Phone() {
    const { phoneNumber } = useParams();
    const title = `Call history for ${phoneNumber}`;
    const [data, isLoading, isError] = useFetchData(apiUrl + phoneNumber);

    const counter = data.reduce((acc, cur) => {
        if (acc[cur.resolution]) {
            acc[cur.resolution]++;
        } else {
            acc[cur.resolution] = 1;
        }
        return acc;
    }, {});

    const chartData = [["Resolution", "Count"], ...Object.entries(counter)];

    return (
        <main>
            {!isLoading && !isError && (
                <>
                    <Table title={title} headers={headers} rows={dataMapper(data)} />
                    <Chart chartType="PieChart" chartData={chartData} />
                </>
            )}
            {isLoading && <h3>Fetching data...</h3>}
            {isError && <h3>Error fetching data. Please refresh your page</h3>}
        </main>
    );
}
export default Phone;
