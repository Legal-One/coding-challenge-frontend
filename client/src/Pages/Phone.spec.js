import React from "react";
import { render } from "@testing-library/react";
import * as mockHook from "../hooks/fetchHook";
import Phone from "./Phone";

jest.mock("react-router-dom", () => ({
    useParams: () => ({
        phoneNumber: "+4912345678",
    }),
}));

describe("PhonePage", () => {
    test("render with all props passed", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [
                {
                    agentName: "John Doe",
                    callTime: "2020-10-06T13:50:00.000Z",
                    resolution: "mock-resolution",
                },
                {
                    agentName: "Spiderman",
                    callTime: "2020-11-06T13:50:00.000Z",
                    resolution: "mock-resolution-2",
                },
            ],
            false,
            false,
        ]);
        const { asFragment } = render(<Phone />);
        expect(asFragment()).toMatchSnapshot();
    });

    test("render with error from api service", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [],
            false,
            true,
        ]);
        const { asFragment } = render(<Phone />);
        expect(asFragment()).toMatchSnapshot();
    });

    test("render when waiting for response from api service", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [],
            true,
            false,
        ]);
        const { asFragment } = render(<Phone />);
        expect(asFragment()).toMatchSnapshot();
    });
});
