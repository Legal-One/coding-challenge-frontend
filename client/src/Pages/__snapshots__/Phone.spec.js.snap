// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhonePage render when waiting for response from api service 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Fetching data...
    </h3>
  </main>
</DocumentFragment>
`;

exports[`PhonePage render with all props passed 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Call history for +4912345678
    </h3>
    <table
      class="Table"
    >
      <thead>
        <tr>
          <th>
            Agent Name
          </th>
          <th>
            Call date and time
          </th>
          <th>
            Resolution
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            John Doe
          </td>
          <td>
            10/6/2020 3:50:00 PM
          </td>
          <td>
            mock-resolution
          </td>
        </tr>
        <tr>
          <td>
            Spiderman
          </td>
          <td>
            11/6/2020 2:50:00 PM
          </td>
          <td>
            mock-resolution-2
          </td>
        </tr>
      </tbody>
    </table>
  </main>
</DocumentFragment>
`;

exports[`PhonePage render with error from api service 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Error fetching data. Please refresh your page
    </h3>
  </main>
</DocumentFragment>
`;
